File: Part 2 - Regression/Section 8 - Decision Tree Regression/R/decision_tree_Regression.R
# Decision Tree REgression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]


# Fitting Decision Tree REgression
library(rpart)
regressor = rpart(formula= Salary~ .,
                  data = dataset,
                  control = rpart.control(minsplit = 3))

#PREDICTING the result
y_pred = predict(regressor, data.frame(Level = 6.5))

#Visualizing the result
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red')+
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            colour = 'blue')+
  ggtitle('Truth or Bluff(Decision Tree Regression)')+
  xlab('Position')+
  ylab('Salary')

File: Part 2 - Regression/Section 8 - Decision Tree Regression/Python/decision_tree_regression.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 29 12:20:47 2020

@author: Dell
"""
                    # Decision Tree Regression
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

dts = pd.read_csv('Position_Salaries.csv')
x = dts.iloc[:, 1:-1].values
y = dts.iloc[:, -1].values

'''Training the Decision Tree Regression'''
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state=0)
regressor.fit(x, y)

## Predicting a new result
regressor.predict([[6.5]])


##Visualising the Decision Tree Regression results (higher resolution)
x_grid = np.arange(min(x), max(x), 0.1)
x_grid = x_grid.reshape(len(x_grid), 1)
plt.scatter(x, y, color='red')
plt.plot(x_grid, regressor.predict(x_grid), color='blue')
plt.title('Truth or Bluff (Support Vector Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()


'''
Decission Tree regression is better when there are several
features'''
'''
There is no need of feature scaling as there is no formulae used 
prediction is just based on splits/distribution'''

